lib/PartitionedProducerImpl.cc

```diff
@@ -266,13 +266,9 @@ void PartitionedProducerImpl::closeAsync(CloseCallback originalCallback) {
             originalCallback(result);
         }
     };
-    if (state_ == Closed) {
-        closeCallback(ResultAlreadyClosed);
-        return;
-    }
-    State expectedState = Ready;
-    if (!state_.compare_exchange_strong(expectedState, Closing)) {
-        return;
+
+    if (state_.exchange(Closed) == Closed) {
+        return closeCallback(ResultAlreadyClosed);
     }
```

Comment:
You should set the state to `Closing`, see comment [here](https://github.com/apache/pulsar-client-cpp/pull/125#discussion_r1033067388)

Reply:
Maybe the following implementation could be better?

```c++
    if (state_ == Closed || state_.exchange(Closing) == Closing) {
        closeCallback(ResultAlreadyClosed);
        return;
    }
```

Reply:
> c++ client state is more than java client and it's more complicated to maintain.

No, the fact is opposite. There are 5 states in C++ client:

```c++
    enum State
    {
        Pending,
        Ready,
        Closing,
        Closed,
        Failed
```

while Java client has 9 states (see `HandlerState.java`):

```java
    enum State {
        Uninitialized, // Not initialized
        Connecting, // Client connecting to broker
        Ready, // Handler is being used
        Closing, // Close cmd has been sent to broker
        Closed, // Broker acked the close
        Terminated, // Topic associated with this handler
                    // has been terminated
        Failed, // Handler is failed
        RegisteringSchema, // Handler is registering schema
        ProducerFenced, // The producer has been fenced by the broker
    }
```

The `Pending` state in C++ client is the `Uninitialized` state in Java client.

> Closing and Closed seems not much different in c++ client.

See the comments in the Java client. `Closing` means the `CloseProducer` command has been sent to the broker and the response is not received yet. In `Closing` state, we should not send any `CloseProducer` command again. It could also be helpful if `producer.close()` stuck. We can debug the application and see the state.